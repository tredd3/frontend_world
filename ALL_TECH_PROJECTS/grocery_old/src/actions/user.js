import {
  SET_USER,
  SET_NEW_AUTH_TOKEN
} from "./constants";
import API_ROUTES from '../helper/network/api-routes';
import {
  fetchApi
} from '../helper/network/fetch';
import store from '../store';
let NewaccessToken;

export const setUser = (data) => {
  return {
    type: SET_USER,
    data: data
  }
}

export const setNewToken = (data) => {
  return {
    type: SET_NEW_AUTH_TOKEN,
    data: data
  }
}

export const addUser = data => {
  return dispatch => {
    return fetchApi({
      url: API_ROUTES.addUser,
      body: data
    }).then(response => {
      dispatch(setUser({ userId: response.Data.userId }));
    }).catch(e => {

    })
    // dispatch(setUser({
    //   userId: "9953"
    // }));
  }
}



export const updateUser = data => {
  return (dispatch, getState) => {
    let nameArr = data.name.split(" ");
    let firstName = nameArr[0];
    const reducer = (accumulator, currentValue, i) => { return (i === 0) ? '' : accumulator + ' ' + currentValue };
    let lastName = nameArr.length !== 1
      ? nameArr.length !== 2
        ? nameArr.reduce(reducer, null)
        : nameArr[1]
      : ""
    fetchApi({
      url: API_ROUTES.updateUser,
      body: {
        userId: getState().user.userId,
        emailId: data.email,
        dob: data.dob,
        gender: data.gender,
        firstName,
        lastName,
        mobileNumber: data.phoneNumber
      }
    })
      .then(response => {
        let user = {
          ...getState().user,
          ...data
        }
        localStorage.setItem('user', JSON.stringify(user));
        return dispatch(setUser({
          ...user
        }));
      })
      .catch(e => {
        console.log("Error", e)
      })
  }
}

export const getUser = data => {
  return dispatch => {
    fetchApi({ url: API_ROUTES.getUserDetail, body: {} })
      .then(response => {
        let webApp = process.env.APP_TYPE == "web" ? true : false;
        let user = {
          customerName: response.Data.customerName,
          firstName: response.Data.firstName,
          lastName: response.Data.lastName,
          userId: response.Data.userId,
          phoneNumber: response.Data.mobileNumber,
          login: true,
          gender: response.Data.gender,
          dob: response.Data.dob,
          email: response.Data.emailId,
        }
        if (webApp) {
          user.access_token = data.access_token;
          user.refresh_token = data.refresh_token;
          user.uuid = data.uuid;
          user.expiresAt = data.expiresAt;
        } else {
          user = { ...store.getState().user, ...user };
        }
        localStorage.setItem('user', JSON.stringify(user));
        return dispatch(setUser({
          ...user
        }));
      })
      .catch(e => {
        let MessageCode = e.Control && e.Control.MessageCode
        if (MessageCode && MessageCode === 404) {
          if (data.history) {
            data.history.push("/account/address/add/addUder");
          }
          // fetchApi({
          //   url: API_ROUTES.addUser,
          //   body: { mobileNumber: "7021143366", firstName: "MyJio", lastName: "User" }
          // }).then(response => {
          //     dispatch(setUser({ userId: response.Data.userId }));
          //   })
          //   .catch(e => {

          //   }) 
        }
      })
  }
}

const addUserToBackend = data => {

}

export const getNewAccessToken = () => {
  return (dispatch, getState) => {
    const AUTH_KEY = process.env.AUTHORISATION_VALUE;
    var params = {
      grant_type: "refresh_token",
      refresh_token: getState().user.refresh_token
    }
    var formBody = [];
    for (var property in params) {
      var encodedKey = encodeURIComponent(property);
      var encodedValue = encodeURIComponent(params[property]);
      formBody.push(encodedKey + "=" + encodedValue);
    }
    formBody = formBody.join("&");
    fetch(API_ROUTES.getNewAccessToken, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        "Authorization": `Basic ${AUTH_KEY}`,
      },
      body: formBody
    }).then(response => response.json())
      .then(response => {
        let userNewTokens = {
          access_token: response.access_token,
          refresh_token: response.refresh_token,
          expiresAt: Date.now() + (response.expires_in * 1000),
        }
        localStorage.setItem('user', JSON.stringify({ ...getState().user, ...userNewTokens }));
        return dispatch(setNewToken(userNewTokens));
      })
      .catch(e => {
        console.log("ERROR HAPPENED During intializing User ")
      })
  }
}
