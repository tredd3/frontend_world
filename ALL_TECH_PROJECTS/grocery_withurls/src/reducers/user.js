import {
  SET_USER, NEW_USER, SET_NEW_AUTH_TOKEN, ORDER_RATED
} from '../actions/constants';

let initialState = {
  // eslint-disable-next-line @typescript-eslint/camelcase
  access_token: null,
  // eslint-disable-next-line @typescript-eslint/camelcase
  refresh_token: null,
  firstName: '',
  lastName: '',
  userId: 0,
  phoneNumber: '',
  login: false,
  xApiKey: null,
  newUser: false,
  newUserErrorMessage: '',
  orderRated: false
};
const placeOrderDataObj = sessionStorage.getItem('placeOrderDataObj');
let sessionUser = null;
if (placeOrderDataObj) {
  sessionUser = JSON.parse(placeOrderDataObj).user;
}
initialState = sessionUser || initialState;

export default function user(state = initialState, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, ...action.data };
    case SET_NEW_AUTH_TOKEN:
      return { ...state, ...action.data };
    case NEW_USER:
      return { ...state, ...action.data };
    case ORDER_RATED:
      return { ...state, ...action.data };
    default:
      return state;
  }
}
