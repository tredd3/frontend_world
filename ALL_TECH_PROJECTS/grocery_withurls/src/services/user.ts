import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { formatDate } from '../helpers/functional';
import { pivotAPIUser } from './helpers/pivots';
import { APIUser, UserPivot } from '../types';
import { genderStringToCode } from './helpers/generic';

export const getUser = promiseMemoize(async (): Promise<UserPivot> => {
  const user = await makeApiCall<{}, APIUser>(apiRoutes.getUserDetail);
  return pivotAPIUser(user);
});

type UpdateUserArgs = {
  userId?: number;
  firstName?: string;
  lastName?: string;
  email?: string;
  gender?: UserPivot['gender'];
  dob?: Date;
  lat?: number;
  lng?: number;
  pincode?: number;
  city?: string;
  storeId?: number;
  storeName?: string;
};

export const updateUser = async ({
  firstName, lastName, email, gender, dob, lat, lng, pincode, city, storeId, storeName
}: UpdateUserArgs) => {
  const userFromApi = await getUser();
  const { userId } = userFromApi;
  const isValidStoreId = storeId !== undefined;
  const isValidStoreName = storeName !== undefined;

  await makeApiCall(apiRoutes.updateUser, {
    body: {
      userId,
      ...(firstName && { firstName }),
      ...(lastName && { lastName }),
      ...(email ? { emailId: email } : { emailId: '' }),
      ...(gender && { gender: genderStringToCode(gender) }),
      ...(dob ? { dob: formatDate(dob) } : { dob: '' }),
      preferences: {
        ...userFromApi.preferences,
        ...(pincode && { pincode }),
        ...(lat && { latitude: lat }),
        ...(lng && { longitude: lng }),
        ...(isValidStoreId && { storeId }),
        ...(city && { city }),
        ...(isValidStoreName && { storeName })
      }
    }
  });

  getUser.clear();
};

type AddUserArgs = {
  firstName: string;
  lastName: string;
};

export const addUser = ({ firstName, lastName }: AddUserArgs) => (
  makeApiCall<AddUserArgs, void>(apiRoutes.addUser, {
    body: {
      firstName,
      lastName
    }
  })
);
