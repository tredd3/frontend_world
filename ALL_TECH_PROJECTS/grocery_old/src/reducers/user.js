import { SET_USER, SET_NEW_AUTH_TOKEN } from '../actions/constants';

let initialState = {
  access_token: null,
  refresh_token: null,
  firstName: "",
  lastName: "",
  userId: 0,
  phoneNumber: '',
  login: false,
  xApiKey: null
}
initialState = JSON.parse(localStorage.getItem('user')) || initialState;

export default function user(state = initialState, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, ...action.data }
    case SET_NEW_AUTH_TOKEN:
      return { ...state, ...action.data }
    default:
      return state;
  }
}
