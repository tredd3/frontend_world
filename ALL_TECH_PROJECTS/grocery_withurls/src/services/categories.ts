import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser } from './user';
import { CategoryPivot } from '../types';

type APICategory = {
  CategoryName: string;
  CategoryId: number;
  CategoryIcon: string;
  CategoryImage: string;
};

type APIRequest = {
  CategoryId?: number;
  Level: 2;
  StoreId: number;
  PageNo: number;
};

type APIResponse = {
  Categories: APICategory[];
};

const convertAPICategoryToCategory = (category: APICategory): CategoryPivot => ({
  id: category.CategoryId,
  name: category.CategoryName,
  icon: category.CategoryIcon,
  image: category.CategoryImage
});

export const getCategories = promiseMemoize(async (id: number) => {
  const { preferences: { storeId } } = await getUser();
  const { Categories } = await makeApiCall<APIRequest, APIResponse>(apiRoutes.getCategories, {
    body: {
      Level: 2,
      ...(id && { CategoryId: id }),
      StoreId: storeId || 0,
      PageNo: 0
    }
  });

  return Categories.map(convertAPICategoryToCategory);
}, { resolve: 'json' });
