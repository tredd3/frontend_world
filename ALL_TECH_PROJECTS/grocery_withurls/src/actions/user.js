import { SET_USER, NEW_USER, SET_NEW_AUTH_TOKEN } from './constants';
import { checkServiceability } from './address';
import { getUser as getUserService, addUser as addUserService } from '../services/user';
import { getConfig } from '../services/config';
import updateUserAction from './update-user';
import { APIError } from '../services/types';

export const updateUser = updateUserAction;

export const setUser = data => ({
  type: SET_USER,
  data
});

export const newUser = data => ({
  type: NEW_USER,
  data
});

export const setNewToken = data => ({
  type: SET_NEW_AUTH_TOKEN,
  data
});

export const setNewUser = () => newUser({ newUser: true });

export const addUser = ({ firstName, lastName }) => async dispatch => {
  try {
    await addUserService({ firstName, lastName });
    getUserService.clear();
    const user = await getUserService();
    dispatch(setUser(user));
    dispatch(newUser({ newUser: false, newUserErrorMessage: '' }));
    try {
      await getConfig();
    } catch (e) {
      console.log(e.message);
    }
  } catch (e) {
    if (e instanceof APIError) {
      dispatch(newUser({ newUserErrorMessage: e.message }));
    }
  }
};

export const getUser = () => async dispatch => {
  try {
    const user = await getUserService();
    return [
      dispatch(setUser(user)),
      dispatch(checkServiceability({ pincode: user.preferences.pincode }))
    ];
  } catch (e) {
    return [
      dispatch(setUser({}))
    ];
  }
};
