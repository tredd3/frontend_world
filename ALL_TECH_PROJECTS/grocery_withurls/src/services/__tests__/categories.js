import mockDB from '../../mocks/db.json';
import { getUser } from '../user';
import { getCategories } from '../categories';

jest.mock('../../intents', () => ({
  getTokens: () => ({
    apiKey: 'dummy-api-key',
    accessToken: 'dummy-access-token'
  })
}));

describe('Categories', () => {
  beforeEach(fetch.resetMocks);
  afterEach(() => {
    getUser.clear();
    getCategories.clear();
  });

  it('should return a properly formatted category', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    fetch.mockResponseOnce(JSON.stringify(mockDB.categories));

    const categories = await getCategories(123);

    expect(JSON.parse(fetch.mock.calls[1][1].body).Data.Request.Body).toMatchSnapshot();
    expect(categories).toMatchSnapshot();
  });
});
