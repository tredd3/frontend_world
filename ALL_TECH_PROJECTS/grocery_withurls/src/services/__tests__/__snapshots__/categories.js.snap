// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories should return a properly formatted category 1`] = `
Object {
  "CategoryId": 123,
  "Level": 2,
  "PageNo": 0,
  "StoreId": 0,
}
`;

exports[`Categories should return a properly formatted category 2`] = `
Array [
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001761_HealthDrinks-Beverages_1574752168.png",
    "id": 10001761,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001761_HealthDrinks-Beverages_1574752168.png",
    "name": "Health Drinks & Beverages",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001760_Beauty-Hygiene_1574752325.png",
    "id": 10001760,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001760_Beauty-Hygiene_1574752325.png",
    "name": "Beauty & Hygiene",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001766_SnacksAndConfectionery_1574752204.png",
    "id": 10001766,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001766_SnacksAndConfectionery_1574752204.png",
    "name": "Snacks And Confectionery",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001765_InstantandReadyFoods_1574752143.png",
    "id": 10001765,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001765_InstantandReadyFoods_1574752143.png",
    "name": "Instant and Ready Foods",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001767_Dairy-Bakery-Egg_1574752358.png",
    "id": 10001767,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001767_Dairy-Bakery-Egg_1574752358.png",
    "name": "Dairy, Bakery & Egg",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001762_Cleaning-HouseholdCare_1574752104.png",
    "id": 10001762,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001762_Cleaning-HouseholdCare_1574752104.png",
    "name": "Cleaning & Household Care",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001763_Foodgrains-OilsAndMasala_1574752241.png",
    "id": 10001763,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001763_Foodgrains-OilsAndMasala_1574752241.png",
    "name": "Foodgrains, Oils And Masala",
  },
  Object {
    "icon": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001764_Mom-Baby_1574752056.png",
    "id": 10001764,
    "image": "https://jiomerchant.akamaized.net/static/category/CustomerOrdImage/10001764_Mom-Baby_1574752056.png",
    "name": "Mom & Baby",
  },
]
`;
